sycl/doc/extensions/SPIRV/SPV_INTEL_float_controls2.asciidoc

```diff
@@ -0,0 +1,265 @@
+
+- Each OpEntryPoint must not set more than one of the *FloatingPointModeALTINTEL* or *FloatingPointModeALTINTEL* execution modes for any given _Target Width_.
+
+Modify the Decoration rules, add the following rules:
+
+- If an *OpFunction* is decorated by more than one *FunctionRoundingModeINTEL* for any given _Target Width_ the _FP Rounding Mode_ operands must match.
+- If an *OpFunction* is decorated by more than one *FunctionDenormModeINTEL* for any given _Target Width_ the _FP Denorm Mode_ operands must match.
+- If an *OpFunction* is decorated by more than one *FunctionFloatingPointModeINTEL* for any given _Target Width_ the _FP Operation Mode_ operands must match.
+
+Modify Section 3.6, Execution Mode:
+
+Add the following rule to the description of *DenormPreserve* execution mode:
+
+If an instruction is inside of a function decorated with *FunctionDenormModeINTEL* with matching _Target Width_ operand, the denorm mode specified by *FunctionDenormModeINTEL* is applied and *DenormPreserve* is ignored.
+
+Add the following rule to the description of *DenormFlushToZero* execution mode:
+
+If an instruction is inside of a function decorated with *FunctionDenormModeINTEL* with matching _Target Width_ operand, the denorm mode specified by *FunctionDenormModeINTEL* is applied and *DenormFlushToZero* is ignored.
+
+Add the following rule to the description of *RoundingModeRTE* execution mode:
+
+If an instruction is not decorated with *FPRoundingMode* and does not define a rounding mode in its description and is inside of a function decorated
+with *FunctionRoundingModeINTEL* with matching _Target Width_ operand, the rounding mode specified by *FunctionRoundingModeINTEL* is applied and *RoundingModeRTE* is ignored.
+
+Add the following rule to the description of *RoundingModeRTZ* execution mode:
+
+If an instruction is not decorated with *FPRoundingMode* and does not define a rounding mode in its description and is inside of a function decorated
+with *FunctionRoundingModeINTEL* with matching _Target Width_ operand, the rounding mode specified by *FunctionRoundingModeINTEL* is applied and *RoundingModeRTZ* is ignored.
+
+Add the following rows to the Execution Mode table:
+--
+[cols="1,20,10,10",options="header",width = "80%"]
+|====
+  2+^| Execution Mode  | Extra Operands | Enabling Capabilities
+| 5620 | *RoundingModeRTPINTEL* +
+The default rounding mode for floating-point arithmetic and conversions instructions must be round to positive infinity.
+If an instruction is decorated with *FPRoundingMode* or defines a rounding mode in its description, that rounding mode is applied and *RoundingModeRTPINTEL* is ignored.
+If an instruction is not decorated with *FPRoundingMode* and does not define a rounding mode in its description and is inside of a function decorated
+with *FunctionRoundingModeINTEL* with matching _Target Width_ operand, the rounding mode specified by *FunctionRoundingModeINTEL* is applied and *RoundingModeRTPINTEL* is ignored. +
```

Comment:
I think we need to define "matching _Target Width_" a little more precisely, especially for operations like **OpFConvert** where there is a source width and a destination width.

Reply:
In this sentence, "matching" was supposed to mean the same _Target Width_ operands in the **RoundingModeRTPINTEL** execution mode and **FunctionRoundingModeINTEL** decoration.

Would the following text be better:

If an instruction is not decorated with **FPRoundingMode** and does not define a rounding mode in its description and is inside of a function decorated with **FunctionRoundingModeINTEL** whose _Target Width_ operand is the same as **RoundingModeRTPINTEL** _Target Width_ operand, the rounding mode specified by **FunctionRoundingModeINTEL** is applied and **RoundingModeRTPINTEL** is ignored. 

Only affects arithmetic instructions operating on a floating-point type whose component width is _Target Width_ and conversion instructions whose result type component width  is _Target Width_.

Reply:
Sorry, I think my previous comment was unclear.

For **OpFConvert** there is the width of the source operand and the width of the result.  Which one of these is checked to determine if the _Target Width_ is matched?

This text kind of makes it seem like it is the source operand that gets checked (due to "operating on"):

> Only affects instructions operating on a floating-point type whose component width is _Target Width_.

That might be a little surprising though, so I'd recommend being a little more precise and explicit.

Reply:
> This text kind of makes it seem like it is the source operand that gets checked (due to "operating on"):
> 
> > Only affects instructions operating on a floating-point type whose component width is _Target Width_.
> 
> That might be a little surprising though, so I'd recommend being a little more precise and explicit.

In my previous comment I proposed an updated text:

Only affects arithmetic instructions operating on a floating-point type whose component width is Target Width and conversion instructions whose result type component width is Target Width.

Reply:
Ah, I think I missed this trying to parse the sentence about the rounding modes.

Do we need to differentiate between arithmetic and conversion instructions, or can we always look at the result type component width to determine if it matches the target width?